libcxx/test/libcxx/ranges/range.adaptors/range.concat/iterator.valueless_by_exception.pass.cpp

```diff
@@ -0,0 +1,265 @@
+  template <std::size_t M>
+  friend struct Iter;
+
+public:
+  Iter() = default;
+  Iter(int* ptr) : ptr_(ptr) {}
+  Iter(const Iter&) = default;
+  Iter(Iter&& other) noexcept : ptr_(other.ptr_) {}
+
+  template <std::size_t M>
+  Iter(const Iter<M>& other) : ptr_(other.ptr_) {}
+
+  Iter& operator=(const Iter&) = default;
+  Iter& operator=(Iter&& other) noexcept {
+    ptr_ = other.ptr_;
+    return *this;
+  }
+
+  reference operator*() const { return *ptr_; }
+  pointer operator->() const { return ptr_; }
+  reference operator[](difference_type n) const { return ptr_[n]; }
+
+  Iter& operator++() {
+    ++ptr_;
+    return *this;
+  }
+  Iter operator++(int) {
+    auto tmp = *this;
+    ++*this;
+    return tmp;
+  }
+  Iter& operator--() {
+    --ptr_;
+    return *this;
+  }
+  Iter operator--(int) {
+    auto tmp = *this;
+    --*this;
+    return tmp;
+  }
+
+  Iter& operator+=(difference_type n) {
+    ptr_ += n;
+    return *this;
+  }
+  Iter& operator-=(difference_type n) {
+    ptr_ -= n;
+    return *this;
+  }
+
+  template <std::size_t X>
+  friend Iter<X> operator+(Iter<X> it, difference_type n);
+
+  template <std::size_t X>
+  friend Iter<X> operator+(difference_type n, Iter<X> it);
+
+  template <std::size_t X>
+  friend Iter<X> operator-(Iter<X> it, difference_type n);
+
+  template <std::size_t X, std::size_t Y>
+  friend difference_type operator-(Iter<X> a, Iter<Y> b);
+
+  friend bool operator==(Iter a, Iter b) = default;
+  friend bool operator<(Iter a, Iter b) { return a.ptr_ < b.ptr_; }
+  friend bool operator>(Iter a, Iter b) { return a.ptr_ > b.ptr_; }
+  friend bool operator<=(Iter a, Iter b) { return a.ptr_ <= b.ptr_; }
+  friend bool operator>=(Iter a, Iter b) { return a.ptr_ >= b.ptr_; }
+};
+
+template <std::size_t X>
+inline Iter<X> operator+(Iter<X> it, std::ptrdiff_t n) {
+  return Iter<X>(it.ptr_ + n);
+}
+
+template <std::size_t X>
+inline Iter<X> operator+(std::ptrdiff_t n, Iter<X> it) {
+  return Iter<X>(it.ptr_ + n);
+}
+
+template <std::size_t X>
+inline Iter<X> operator-(Iter<X> it, std::ptrdiff_t n) {
+  return Iter<X>(it.ptr_ - n);
+}
+
+template <std::size_t X, std::size_t Y>
+inline std::ptrdiff_t operator-(Iter<X> a, Iter<Y> b) {
+  return a.ptr_ - b.ptr_;
+}
+
+template <std::size_t N>
+struct Range : std::ranges::view_base {
+  using iterator       = Iter<N>;
+  using const_iterator = Iter<N>;
+
+  int* data_;
+  std::size_t size_;
+
+  Range() : data_(val), size_(4) {}
+
+  Range(int* data, std::size_t size) : data_(data), size_(size) {}
+
+  iterator begin() { return iterator(data_); }
+  iterator end() { return iterator(data_ + size_); }
+
+  const_iterator begin() const { return const_iterator(data_); }
+  const_iterator end() const { return const_iterator(data_ + size_); }
+
+  std::size_t size() const { return size_; }
+};
+
+static_assert(std::ranges::range<Range<0>>);
+static_assert(std::ranges::sized_range<Range<0>>);
+
+int main() {
+  {
+    //valueless by exception test operator*
```

Comment:
For usual comments, it's conventional to have a space after `//`. Ditto below.
```suggestion
    // valueless by exception test operator*
```